renormalizer/model/basis.py

```diff
@@ -424,58 +436,67 @@ def __init__(self, dof, nbas, xi, xf, endpoint=False):
     def __str__(self):
         return f"BasisSineDVR(xi: {self.xi}, xf: {self.xf}, nbas: {self.nbas})"
 
     def op_mat(self, op: Union[Op, str]):
+        
         if not isinstance(op, Op):
             op = Op(op, None)
         op_symbol, op_factor = op.symbol, op.factor
-
+        
+        # partialx is deprecated
+        op_symbol = op_symbol.replace("partialx", "dx")
+        
+        # operators having analytical matrix elements
         if op_symbol == "I":
             mat = np.eye(self.nbas)
 
-        elif op_symbol.split("^")[0] == "x" and " " not in op_symbol:
-            if len(op_symbol.split("^")) == 1:
-                # legacy for check
-                mat1 = np.zeros((self.nbas, self.nbas))
-                for j in range(1,self.nbas+1,1):
-                    for k in range(1,self.nbas+1,1):
-                        a1 = (j+k)*np.pi/self.L
-                        a2 = (j-k)*np.pi/self.L
-                        if (j+k)%2 == 1:
-                            res = -1/self.L*(-2/a1**2+2/a2**2)
-                        elif j-k == 0:
-                            res = self.xi + 0.5*self.L
-                        else:
-                            res = 0
-                        mat1[j-1,k-1] = res
-
-                mat = self._I()*self.xi+self._u()
-                assert np.allclose(mat, mat1)
+        elif op_symbol == "x":
+            # legacy for check
+            mat1 = np.zeros((self.nbas, self.nbas))
+            for j in range(1,self.nbas+1,1):
+                for k in range(1,self.nbas+1,1):
+                    a1 = (j+k)*np.pi/self.L
+                    a2 = (j-k)*np.pi/self.L
+                    if (j+k)%2 == 1:
+                        res = -1/self.L*(-2/a1**2+2/a2**2)
+                    elif j-k == 0:
+                        res = self.xi + 0.5*self.L
+                    else:
+                        res = 0
+                    mat1[j-1,k-1] = res
 
-                mat = self.dvr_v.T @ mat @ self.dvr_v
-            else:
-                moment = float(op_symbol.split("^")[1])
-                if moment == 1:
-                    mat = self.op_mat("x")
-                if moment == 2:
-                    mat = self._I()*self.xi**2+self._u()*self.xi*2+self._uu()
-                    mat = self.dvr_v.T @ mat @ self.dvr_v
-                else:
-                    mat = np.diag(self.dvr_x ** moment)
+            mat = self._I()*self.xi+self._u()
+            assert np.allclose(mat, mat1)
 
+        elif op_symbol == "x^1":
+            self._recursion_flag = 1
+            mat = self.op_mat("x")
+
+        elif op_symbol == "x^2":
+             mat = self._I()*self.xi**2+self._u()*self.xi*2+self._uu()
+        
+        elif op_symbol == "x^3":
+            mat = self._I()*self.xi**3 + 3*self._uu()*self.xi + 3*self._u()*self.xi**2 + self._uuu()
+        
         elif set(op_symbol.split(" ")) == set("x"):
             moment = len(op_symbol.split(" "))
+            self._recursion_flag = 1
```

Comment:
Is it better to use `self._recursion_flag += 1` just as BasisSHO for deep recursion?

Reply:
ok. I will do it.

Reply:
solved